src/java/org/apache/cassandra/config/DatabaseDescriptor.java

```diff
@@ -4763,4 +4756,159 @@ public static OptionalDouble getSeverityDuringDecommission()
                OptionalDouble.of(conf.severity_during_decommission) :
                OptionalDouble.empty();
     }
+
+    public static void visit(PropertyVisitor visitor)
+    {
+        visit(visitor, RuntimeException::new);
+    }
+
+    public static void visit(PropertyVisitor visitor, Function<Throwable, ? extends RuntimeException> handler)
+    {
+        try
+        {
+            for (Property property : ofNullable(confValueAccessors).orElse(Collections.emptyMap()).values())
+                visitor.visit(property.getName(), property.getType(), property.getAnnotation(Mutable.class) == null);
+        }
+        catch (Throwable t)
+        {
+            throw handler.apply(t);
+        }
+    }
+
+    public static <T> ListenableProperty.Remover addBeforeChangePropertyListener(String name, Class<T> clazz, BiConsumer<T, T> consumer)
+    {
+        return addPropertyListener(name, clazz, p -> p.addBeforeHandler(ListenableProperty.Handler.consume(consumer)));
+    }
+
+    public static <T> ListenableProperty.Remover addAfterChangePropertyListener(String name, Class<T> clazz, BiConsumer<T, T> consumer)
+    {
+        return addPropertyListener(name, clazz, p -> p.addAfterHandler(ListenableProperty.Handler.consume(consumer)));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> ListenableProperty.Remover addPropertyListener(String name,
+                                                                      Class<T> clazz,
+                                                                      Function<ListenableProperty<Config, T>, ListenableProperty.Remover> adder)
+    {
+        Property property = ofNullable(confValueAccessors.get(name))
+                            .map(p -> {
+                                // The case when replacement property is used and match the property's name.
+                                if (p instanceof Replacement.ReplacementProperty)
+                                    return ((Replacement.ReplacementProperty) p).delegate();
```

Comment:
this is a bug, why are you unwrapping?  If I want to listen to a old property why are you blocking me?

Reply:
I think we should not be able to listen to any replacements changes for the following reasons:
- there are few replacements with the same name, but their type has been replaced; For example, key_cache_save_period, row_cache_save_period, counter_cache_save_period. This means that both of these types a new one and the replaced one will be listenable, there is no way to pick up the correct property to listen to. We end up with a listener type that doesn't match the property type, which sounds like a bad design;
- Replacements exist only for users and public API (Yaml, SettingsTable). In fact, internally all the code that depends on them has been migrated on that was migrated to new properties; There is no need to support it;
-  most of the Replacements are deprecated: 86 out of 92. 

So, as we discussed privately let's skip the Replacements for now.